File: Problem3.java
// Time Complexity : push(x) - O(n) pop() - O(1) peek() - O(1) empty() - O(1)
// Space Complexity : O(n) [O(2n)]
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : I am not sure if this is the efficient way or not.


// Your code here along with comments explaining your approach

class MyQueue {
    
    private Stack<Integer> st1; //Stack for storing values
    private Stack<Integer> st2; //Used as buffer to store the elements in reverse order
    int size;

    public MyQueue() {
        //Initialising values
        st1 = new Stack<Integer>();
        st2 = new Stack<Integer>();
        size = 0;
    }
    
    public void push(int x) {
        while(!st1.empty()){ //transfering all elements to stack2
            st2.push(st1.pop()); 
        }
        st1.push(x); // adding element to stack1
        size++;
        while(!st2.empty()){ //transfering all elements back to stack1
            st1.push(st2.pop());
        }
    }
    
    public int pop() {
        size--;
        return st1.pop(); //pop the top element
    }
    
    public int peek() {
        return st1.peek(); //peek the top element
    }
    
    public boolean empty() {
        return st1.empty(); //checking whether the queue is empty
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * boolean param_4 = obj.empty();
 */

File: Problem3.py
class MinStack:
    #Time Complexity :O(1)
    #Space Complexity :O(n)
    #Did this code successfully run on Leetcode : Yes
    #Any problem you faced while coding this : No

    #there are 2 approaches with which can solve this problem:
    #1-> using 2 stacks : main stack and min stack, pushing all values to the main stack and minimum values in min stack for each value.We will be Poping values from both stacks. 
    #2-> using 1 stack: pushing the values to the stack. if the new value is less than or equal to previous minimum value, pushing the previous minimum value before the new value. Poping:if the poped value is equal to the minimum value, pop the previous value and set it to minimum.  
    def __init__(self):
        self.stack=[]
        self.min=float(inf)

    def push(self, val: int) -> None:
        if val<=self.min:
            self.stack.append(self.min)
            self.min=val
        self.stack.append(val) 

    def pop(self) -> None:
        if self.min==self.stack.pop():
            self.min=self.stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()